File: menubar/Sources/WindowArrangerCore/Who.swift
import Foundation

// Swift port of lib/who.ts — the FROZEN /api/who v1.4 contract. census polls it
// at 1 Hz and republishes it PUBLICLY: field names, nesting, and the idle privacy
// clamp are load-bearing. Wire key order matches the TS return literal:
// ts, fleet, windows, displays, spaces, profile, idle.
public enum Who {
    public static let shells: Set<String> = ["zsh", "sh", "-zsh", "bash", "-bash"]

    public static func report(snapshot s: Arranger.SnapshotData) -> JSON {
        let names = Displays.names()
        let pins = Pins.load()

        func windowDisplay(_ w: YabaiWindow) -> Int? {
            w.display ?? s.displays.first { $0.spaces.contains(w.space) }?.index
        }
        func displayName(forIndex idx: Int) -> String? {
            guard let d = s.displays.first(where: { $0.index == idx }) else { return nil }
            return names[String(d.id)] ?? "display \(d.index)"
        }

        // fleet — WezTerm roster.
        struct FleetSeed {
            let title: String; let id: Int; let display: Int; let space: Int; let focus: Bool
        }
        let fleetSeeds: [FleetSeed] = s.windows.compactMap { w in
            guard w.app == "WezTerm" else { return nil }
            let title = w.title.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !title.isEmpty, let disp = windowDisplay(w) else { return nil }
            return FleetSeed(title: title, id: w.id, display: disp, space: w.space, focus: w.hasFocus)
        }
        var dupCount: [String: Int] = [:]
        for f in fleetSeeds { dupCount[f.title.lowercased(), default: 0] += 1 }
        let fleet: [JSON] = fleetSeeds.map { f in
            .obj([
                ("title", .s(f.title)),
                ("app", .s("WezTerm")),
                ("id", .i(f.id)),
                ("display", .i(f.display)),
                ("space", .i(f.space)),
                ("focus", .b(f.focus)),
                ("dup", .i(dupCount[f.title.lowercased()] ?? 1)),
                ("displayName", displayName(forIndex: f.display).map(JSON.s) ?? .null),
                ("isShell", .b(shells.contains(f.title.lowercased()))),
            ])
        }

        // windows — the v1.2 mini-map: all apps, all spaces, real frames only.
        let windows: [JSON] = s.windows.compactMap { w in
            let title = w.title.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !title.isEmpty, let disp = windowDisplay(w),
                  w.frame.w >= 80, w.frame.h >= 60 else { return nil }
            let pinned = pins.contains { Pins.matches(w, $0) }
            let idleOnly = pins.contains { $0.whenIdleOnly && Pins.matches(w, $0) }
            return .obj([
                ("id", .i(w.id)),
                ("app", .s(w.app)),
                ("title", .s(title)),
                ("space", .i(w.space)),
                ("display", .i(disp)),
                ("focus", .b(w.hasFocus)),
                ("pinned", .b(pinned)),
                ("whenIdleOnly", .b(idleOnly)),
                ("frame", .obj([("x", .n(w.frame.x)), ("y", .n(w.frame.y)),
                                ("w", .n(w.frame.w)), ("h", .n(w.frame.h))])),
            ])
        }

        let displays: [JSON] = s.displays.map { d in
            .obj([
                ("index", .i(d.index)),
                ("name", .s(names[String(d.id)] ?? "display \(d.index)")),
                ("frame", .obj([("x", .n(d.frame.x)), ("y", .n(d.frame.y)),
                                ("w", .n(d.frame.w)), ("h", .n(d.frame.h))])),
            ])
        }

        let spaces: [JSON] = s.spaces.map { sp in
            let pinned = pins.contains { p in
                p.space == sp.index || (p.label != nil && p.label == sp.label)
            }
            return .obj([
                ("index", .i(sp.index)),
                ("display", .i(sp.display)),
                ("isVisible", .b(sp.isVisible)),
                ("hasFocus", .b(sp.hasFocus)),
                ("pinned", .b(pinned)),
            ])
        }

        let profile = profileStatus(fleetSeeds: fleetSeeds.map { ($0.title, $0.space) },
                                    spaces: s.spaces, pins: pins)

        // idle — privacy-clamped for the public wall: 10s grain, capped at the
        // threshold (a live "seconds since Nat touched input" is a burglar clock).
        // `armed` uses the TRUE unclamped value.
        let idleSecs = Idle.effectiveSeconds()
        let clamped: JSON = idleSecs.map { .n(min(Idle.thresholdSec, ($0 / 10).rounded(.down) * 10)) } ?? .null
        let armed = idleSecs != nil && idleSecs! >= Idle.thresholdSec
            && pins.contains { $0.whenIdleOnly }
        let idle: JSON = .obj([
            ("seconds", clamped),
            ("threshold", .n(Idle.thresholdSec)),
            ("visibleThreshold", .n(Idle.thresholdSec * Idle.visibleMultiplier)),
            ("armed", .b(armed)),
        ])

        return .obj([
            ("ts", .n(s.ts)),
            ("fleet", .arr(fleet)),
            ("windows", .arr(windows)),
            ("displays", .arr(displays)),
            ("spaces", .arr(spaces)),
            ("profile", profile),
            ("idle", idle),
        ])
    }

    // {stale, reason?} | null. Two stale signals, reasons joined with "; ":
    // lost pin labels (synthetic ⭐ ones excluded) and ≥2 oracles off their profile
    // space (one drifter is tolerated — it's indistinguishable from an intent).
    static func profileStatus(fleetSeeds: [(title: String, space: Int)],
                              spaces: [YabaiSpace], pins: [PinRule]) -> JSON {
        let profile = OracleProfile.load()
        guard !profile.isEmpty else { return .null }
        var reasons: [String] = []

        let liveLabels = Set(spaces.compactMap { $0.label }.filter { !$0.isEmpty })
        let lost = pins.compactMap { $0.label }
            .filter { !liveLabels.contains($0) && !Pins.isSyntheticLabel($0) }
        var seenLost = Set<String>()
        let uniqueLost = lost.filter { seenLost.insert($0).inserted }
        if !uniqueLost.isEmpty {
            reasons.append("pin label(s) lost after reboot: \(uniqueLost.joined(separator: ", "))")
        }

        var mismatched = 0
        for f in fleetSeeds where !shells.contains(f.title.lowercased()) {
            let t = f.title.lowercased()
            // Oracle titles are single tokens (repo names); a spaced title is a
            // status pane ("2 awaiting input · claude agents", "Copy mode: …")
            // whose substring can steal a real rule (…awai-TING…) and fake a drift.
            if t.contains(" ") { continue }
            guard let rule = profile.first(where: { r in
                let m = r.match.lowercased()
                // JS includes("") is true — Pins.has replicates that.
                return Pins.has(t, m) || Pins.has(m, t)
            }), let want = rule.space else { continue }
            if want != f.space { mismatched += 1 }
        }
        if mismatched >= 2 {
            reasons.append("\(mismatched) oracle(s) off their profile space (likely space renumber)")
        }

        if reasons.isEmpty { return .obj([("stale", .b(false))]) }
        return .obj([("stale", .b(true)), ("reason", .s(reasons.joined(separator: "; ")))])
    }
}

File: lib/who.ts
// who — shared fleet-topology computation: which oracle is on which display/space.
// Extracted from scripts/who.ts so the CLI and the HTTP endpoint (server.ts
// GET /api/who) share one implementation instead of duplicating the shape.
import type { Snapshot, YabaiSpace } from "./bsp";
import { displayNames } from "./displays";
import { loadProfile, loadPins, pinMatches, isSyntheticLabel, type OracleRule } from "./oracle-profile";
import { effectiveIdleSeconds, IDLE_THRESHOLD_SEC, IDLE_VISIBLE_MULTIPLIER } from "./idle";

const SHELLS = new Set(["zsh", "sh", "-zsh", "bash", "-bash"]);

// Is the saved profile stale vs the live layout? Two reboot-caused signals:
//  1. a pin's space LABEL no longer resolves (yabai labels don't survive reboot,
//     so the pin is silently disarmed).
//  2. open oracles sit on a different space than their profile rule (space
//     indices renumber on reboot / display reshuffle → raw-index rules drift).
// Returns null when there's no profile at all (→ display-census shows no badge).
function computeProfileStatus(
  fleet: { title: string; space: number; isShell: boolean }[],
  spaces: YabaiSpace[],
): { stale: boolean; reason?: string } | null {
  const profile = loadProfile();
  if (!profile.length) return null;

  const matchRule = (title: string): OracleRule | undefined => {
    const t = title.toLowerCase();
    return profile.find((r) => t.includes(r.match.toLowerCase()) || r.match.toLowerCase().includes(t));
  };

  const reasons: string[] = [];

  const labels = new Set(spaces.map((s) => s.label).filter((l): l is string => !!l));
  const lostLabels = [...new Set(loadPins().map((p) => p.label).filter((l): l is string => !!l))]
    // Synthetic ⭐ labels (auto-assigned by space-pin) self-heal within 60s and
    // are never user-named — keep them out of the census-facing stale reason.
    .filter((l) => !labels.has(l) && !isSyntheticLabel(l));
  if (lostLabels.length) reasons.push(`pin label(s) lost after reboot: ${lostLabels.join(", ")}`);

  let mismatched = 0;
  for (const w of fleet) {
    if (w.isShell) continue;
    // Oracle titles are single tokens (repo names); a spaced title is a status
    // pane ("2 awaiting input · claude agents", "Copy mode: …") whose substring
    // can steal a real rule (…awai-TING…) and fake a drift.
    if (w.title.includes(" ")) continue;
    const rule = matchRule(w.title);
    if (rule?.space != null && rule.space !== w.space) mismatched++;
  }
  if (mismatched >= 2) reasons.push(`${mismatched} oracle(s) off their profile space (likely space renumber)`);

  return reasons.length ? { stale: true, reason: reasons.join("; ") } : { stale: false };
}

export interface FleetRow {
  title: string;
  app: string;
  id: number;
  display: number;
  space: number;
  focus: boolean;
  dup: number;
  displayName: string | null;
  isShell: boolean;
}

// display-census schema v1.2 (additive): EVERY window on EVERY space — feeds
// the per-space mini-map on the public site. frame is yabai's raw GLOBAL
// coordinates (consumer normalizes against its display's frame).
export interface WindowRow {
  id: number;
  app: string;
  title: string;
  space: number;
  display: number;
  focus: boolean;
  // v1.3 (additive): this window matches an ACTIVE pin rule (live enforcement
  // state — false for everything while the global pin toggle is off).
  pinned: boolean;
  // v1.4 (additive): this window matches a whenIdleOnly pin — it parks to its
  // rule's target automatically once the human idles past the threshold.
  whenIdleOnly: boolean;
  frame: { x: number; y: number; w: number; h: number };
}

export interface WhoReport {
  ts: number;
  fleet: FleetRow[];
  displays: { index: number; name: string; frame: { x: number; y: number; w: number; h: number } }[];
  // v1.3 (additive): spaces[].pinned — some active pin rule targets this space.
  spaces: { index: number; display: number; isVisible: boolean; hasFocus: boolean; pinned: boolean }[];
  windows: WindowRow[];
  // display-census schema v1.1: amber "⚠ profile stale" badge when stale=true.
  profile: { stale: boolean; reason?: string } | null;
  // v1.4 (additive): the shared fleet idle clock (census 🌙 badge). seconds is
  // PRIVACY-CLAMPED — /api/who is republished on the PUBLIC census site, and a
  // live "seconds since Nat touched input" counter is a burglar clock, so it's
  // quantized to 10s and capped at the threshold (armed still uses the true
  // value). null = idle binary unavailable. armed = some whenIdleOnly pin is
  // active AND the base threshold passed (windows on a visible space wait for
  // visibleThreshold — see enforcePins).
  idle: { seconds: number | null; threshold: number; visibleThreshold: number; armed: boolean };
}

export async function computeWhoReport(snapshot: Snapshot): Promise<WhoReport> {
  const names = await displayNames();
  const nameOf = (d: { id: number; index: number }) => names[String(d.id)] ?? `display ${d.index}`;

  const rows = snapshot.windows
    .filter((w) => w.app === "WezTerm" && (w.title ?? "").trim() && w.space != null && w.display != null)
    .map((w) => ({
      title: (w.title ?? "").trim(),
      app: w.app,
      id: w.id,
      display: w.display!,
      space: w.space!,
      focus: !!(w as any)["has-focus"],
      dup: 0,
    }));

  const counts = new Map<string, number>();
  for (const r of rows) counts.set(r.title.toLowerCase(), (counts.get(r.title.toLowerCase()) ?? 0) + 1);
  for (const r of rows) r.dup = counts.get(r.title.toLowerCase())!;

  const fleet: FleetRow[] = rows.map((r) => {
    const d = snapshot.displays.find((x) => x.index === r.display);
    return { ...r, displayName: d ? nameOf(d) : null, isShell: SHELLS.has(r.title.toLowerCase()) };
  });

  // v1.2 mini-map rows: all apps, all spaces (invisible ones included — the
  // snapshot already has them). Noise filter per display-census: drop tiny
  // chrome (w<80 or h<60) and untitled windows; KEEP minimized ones with a
  // real frame so the map shows them where they'd restore.
  const pins = loadPins(); // [] while the global toggle is off → pinned:false everywhere (truthful)
  const windows: WindowRow[] = snapshot.windows
    .filter((w) => w.space != null && w.display != null && (w.title ?? "").trim() !== "" &&
      w.frame != null && w.frame.w >= 80 && w.frame.h >= 60)
    .map((w) => ({
      id: w.id,
      app: w.app,
      title: (w.title ?? "").trim(),
      space: w.space!,
      display: w.display!,
      focus: !!(w as any)["has-focus"],
      pinned: pins.some((p) => pinMatches(w, p)),
      whenIdleOnly: pins.some((p) => p.whenIdleOnly && pinMatches(w, p)),
      frame: { x: w.frame!.x, y: w.frame!.y, w: w.frame!.w, h: w.frame!.h },
    }));

  const idleSecs = effectiveIdleSeconds();
  return {
    ts: snapshot.ts,
    fleet,
    windows,
    displays: snapshot.displays.map((d) => ({ index: d.index, name: nameOf(d), frame: d.frame })),
    spaces: snapshot.spaces.map((s) => ({
      index: s.index,
      display: s.display,
      isVisible: !!(s as any)["is-visible"],
      hasFocus: !!(s as any)["has-focus"],
      pinned: pins.some((p) => p.space === s.index || (p.label != null && p.label === s.label)),
    })),
    profile: computeProfileStatus(fleet, snapshot.spaces),
    idle: {
      seconds: idleSecs == null ? null : Math.min(IDLE_THRESHOLD_SEC, Math.floor(idleSecs / 10) * 10),
      threshold: IDLE_THRESHOLD_SEC,
      visibleThreshold: IDLE_THRESHOLD_SEC * IDLE_VISIBLE_MULTIPLIER,
      armed: idleSecs != null && idleSecs >= IDLE_THRESHOLD_SEC && pins.some((p) => p.whenIdleOnly),
    },
  };
}
